File: app/services/location_finder.rb
require "uri"
require "net/http"

class LocationFinder < ApplicationService
  attr_reader :location
  PLACES_API = "https://maps.googleapis.com/maps/api/place/textsearch/json?"

  def initialize(location, current_location)
    @location = location
    @current_location = current_location
  end

  def call
    params = {
      query: @location,
      fields: "formatted_address,name,geometry",
      location: @current_location,
      key: ENV["GOOGLE_MAPS_KEY"]
    }
    url = URI(PLACES_API+params.to_param)

    https = Net::HTTP.new(url.host, url.port)
    https.use_ssl = true

    request = Net::HTTP::Get.new(url)

    response = https.request(request)
    location = Locations.new
    location.from_json(response.read_body)
    list = location.results.map do |result|
      LocationResult.new(result["formatted_address"], result["geometry"], result["name"])
    end
    return list
  end
end

File: lib/location-finder.ts
import { Locations, LocationResult } from './models';

export class LocationFinder {
  private location: string;
  private currentLocation?: string;

  constructor(location: string, currentLocation?: string) {
    this.location = location;
    this.currentLocation = currentLocation;
  }

  async call(): Promise<LocationResult[]> {
    const placesApi =
      process.env.PLACES_API ||
      'https://maps.googleapis.com/maps/api/place/textsearch/json?';

    const params = new URLSearchParams({
      query: this.location,
      fields: 'formatted_address,name,geometry',
      key: process.env.GOOGLE_MAPS_KEY || '',
    });

    // Only add location bias if currentLocation is provided
    if (this.currentLocation) {
      params.append('location', this.currentLocation);
    }

    const url = `${placesApi}${params.toString()}`;
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`Google Places API error: ${response.statusText}`);
    }

    const data: Locations = await response.json();

    if (data.status !== 'OK' && data.status !== 'ZERO_RESULTS') {
      throw new Error(
        data.error_message || `API returned status: ${data.status}`
      );
    }

    return data.results.map((result) => ({
      formatted_address: result.formatted_address,
      geometry: result.geometry,
      name: result.name,
    }));
  }

  static async call(
    location: string,
    currentLocation?: string
  ): Promise<LocationResult[]> {
    const finder = new LocationFinder(location, currentLocation);
    return finder.call();
  }
}

